File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.java
class Solution {
    public int minSetSize(int[] arr) {
        Map<Integer,Integer> counter = new HashMap<Integer,Integer>();
        
        for(int num : arr) {
            counter.put(num,counter.getOrDefault(num,0) + 1);
        }
        
        Integer[] counts = counter.values().toArray(new Integer[counter.size()]);
        
        Arrays.sort(counts,(a, b) -> b - a);
        
        long sum = 0;
        int res = 0;
        for(int c : counts) {
            
            res++;
            sum += c;
            if(sum >= arr.length / 2) break;
        }
        
        return res;
    }
}

File: 1338-reduce-array-size-to-the-half/1338-reduce-array-size-to-the-half.js
/**
 * @param {number[]} arr
 * @return {number}
 */
var minSetSize = function(arr) {
    const map = new Map();
    const heap = new MaxPriorityQueue({compare: (a,b) => {
        return b[1] - a[1];
    }});
    
    for(const a of arr) {
        const value = map.get(a) | 0;
        map.set(a,value + 1);
    }
    
    for(const pair of map) {
        heap.enqueue(pair);
    }
    
    let sum = 0;
    let res = 0;
    
    do {
        sum += heap.dequeue()[1];
        res++;
    }while(sum < arr.length / 2);
    
    return res;
};